README
NAME
    Data::Embed - embed arbitrary data in a file

VERSION
    version 0.21

SYNOPSIS
       use Data::Embed qw< embed embedded >;

       # this is the file where thing will be embedded, at the end
       my $container = '/path/to/some/file';

       # first of all we embed an external file
       my $datafile  = '/path/to/data.tar.gz';
       embed($container, name => 'data.tar.gz', filename => $datafile);

       # we can also embed some data, directly
       use Data::Dumper;
       my $conf = { ... };
       embed($container, name => 'config.yml', data => Dumper($conf));

       # if the data is in a scalar but it's huge, use filename and
       # pass a reference to the scalar so no copy will happen
       my $huge_png = ...;
       embed($container, name => 'image.png', filename => \$huge_png);

       # to retrieve the stuff, use embedded()
       my @files = embedded($container);

       # each item in @files is a Data::Embed::File object

       # get whole contents of file
       my $config_text = $files[1]->contents();

       # otherwise, you can get a filehandle and use it, e.g. to
       # dump it on standard output
       my $data_fh = $files[0]->fh();
       binmode STDOUT;
       print {*STDOUT} <$data_fh>;

       # or save the file back, using the available name
       open my $ofh, '>:raw', $file[2]->name(); # well, do your checks!
       my $ifh = $files[2]->fh();
       while (! eof $ifh) {
          read $ifh, my $buffer, 4096
             or last; # do proper checks in production!
          print {$ofh} $buffer;
       }

DESCRIPTION
    This module allows you to manage embedding data at the end of other
    files, providing both means for embedding the data ("embed" and
    "writer") and accessing them ("embedded" and "reader").

    How can this be helpful? For example, suppose that you want to bring
    some data along with your perl script, some of which might be binary
    (e.g. an image, or a tar archive), you can embed these data inside the
    perl and then retrieve them. For example, this can be the basis for an
    installer.

    For embedding data, you can use the "embed" function. See the relevant
    documentation or the examples in the "SYNOPSYS" to use it properly.

    For extracting the embedded data, you can use the "embedded" function
    and access each embedded file as a Data::Embed::File object. You can
    then use its methods "contents" for accessing the whole data, or get a
    filehandle through "fh" and avoid getting the whole data in memory at
    once.

    Note: the filehandle provided by the "fh" method of Data::Embed::File is
    actually a IO::Slice object, so it might not support all the
    functions/methods of a regular filehandle.

    You can also access the lower level interface through the two functions
    "reader" and "writer". See the documentation for Data::Embed::Reader and
    Data::Embed::Writer.

FUNCTIONS
  embed
       embed($hashref); # OR
       embed(%keyvalue_pairs);

    Embed new data inside a container file.

    Parameters can be passed as key-value pairs either directly or through a
    hash reference. The following keys are supported:

    "container"
        shortcut to specifying the same input and output, i.e. the value
        will be replicated both on the "input" and "output" keys below.
        Caller still has to ensure that the two are compatible. Provision of
        a filehandle is currently not supported.

    "input"
        any previous container file to use as base for the generated
        container. If missing, no previous data will be considered (like
        starting from an empty file). Can be:

        *   the "-" string in a plain scalar, in which case standard input
            is considered

        *   any other string in a plain scalar, considered to be a file name

        *   a plain reference to a scalar, considered to hold the input data

        *   something that supports the filehandle interface for reading

    "output"
        the target container for the newly generated archive. Might be the
        same as the input or different; in the latter case, the input will
        be copied over the output, apart from the bits regarding the
        management of the inclusions. Can be:

        *   missing/undefined or the "-" string in a plain scalar, in which
            case standard output is used

        *   any other string in a plain scalar, considered to be a file name

        *   a plain reference to a scalar, considered to be the target
            scalar to hold the data

        *   something that supports the filehandle interface for printing.
            You should not provide the same filehandle for both input and
            output, even if you opened it in read-write mode. This
            limitation might be removed in the future.

    "name"
        the name to associate to the section, optionally. If missing it will
        be set to the empty string

    "fh"
        the filehandle from where data should be taken. The filehandle will
        be exausted starting from its current position

    "filename"
        a filename or a reference to a scalar where data will be read from

    "data"
        a scalar from where data will be read. If you have a huge amount of
        data, it's better to use the "filename" key above passing a
        reference to the scalar holding the data.

    Options "fh", "filename" and "data" are exclusive and will be considered
    in the order above (first come, first served).

    This function does not return anything.

  embedded
    Get a list of the embedded files inside a target container. The calling
    syntax is as follows:

       my $arrayref = embedded($container); # scalar context, OR
       my @files    = embedded($container); # list context

    The only input parameter is the $container to use as input. It can be
    either a real filename, or a filehandle.

    Depending on the context, a list will be returned (in list context) or
    an array reference holding the list.

    Whatever the context, each item in the list is a Data::Embed::File
    object that you can use to access the embedded file data (most notably,
    you'll be probably using its "contents" or "fh" methods).

  writer
    This is a convenience wrapper around the constructor for
    Data::Embed::Writer.

  reader
    This is a convenience wrapper around the constructor for
    Data::Embed::Reader.

  generate_module_from_file
       # when %args includes details for an output channel
       generate_module_from_file(%args);

       # in case no output is provided in %args:
       my $text = generate_module_from_file(%args);

    Generate a module's file contents from a file. The module contains code
    of a package that has code to read the included data. Arguments are:

    package
        the name of the package that will be put into the module. This is a
        mandatory parameter.

    output
        the output channel. If not present, the output will be provided as a
        string returned by the function, otherwise you can provide

        *   a filehandle where the output will be printed

        *   a reference to a scalar (it will be filled with the contents)

        *   the "-" string, in which case the output will be printed to
            STDOUT

        *   a filename

    output_from_package
        if this key is present and true, the "output" parameters is
        overridden and generated automatically from the package name
        provided in key "package". The generated file will assume that the
        file is contained in the *normal* path under a "lib" directory, e.g.
        if the package name is "Some::Module" then the generated filename
        will be "lib/Some/Module.pm".

    fh  a filehandle where data will be read from

    filename
        the input will be taken from the provided filename

    dataref
        the input will be taken from the scalar pointed by the reference

    data
        the input is taken from the scalar provided with the data key

    Input keys are "fh", "filename", "dataref" and "data". In case multiple
    of them are present, they will be considered in the order specified.

BUGS AND LIMITATIONS
    Report bugs either through RT or GitHub (patches welcome).

    Passing the same filehandle for both "input" and "output" in "embed" is
    not supported. This applies to "container" too.

SEE ALSO
    Data::Section covers a somehow similar need but differently. In
    particular, you should look at it if you want to be able to modify the
    data you want to embed directly, e.g. if you are embedding some textual
    templates that you want to tweak.

AUTHOR
    Flavio Poletti <polettix@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2014-2015 by Flavio Poletti <polettix@cpan.org>

    This module is free software. You can redistribute it and/or modify it
    under the terms of the Artistic License 2.0.

    This program is distributed in the hope that it will be useful, but
    without any warranty; without even the implied warranty of
    merchantability or fitness for a particular purpose.



Changes
0.21      2015-01-06 19:43:19 Europe/Rome
   - release after 64 passes on CPAN Testers (thanks!)

0.2_03    2015-01-06 07:32:02 Europe/Rome
   - fixed test t/10-writer-add.t for openining a handle to a scalar
     ref in perl prior to 5.14 (no ':raw' allowed, using binmode()
     just after)

0.2_02    2015-01-06 01:38:39 Europe/Rome
   - new D::E::Writer and writer() implementation (sanitized)
   - added documentation for data-embed script
   - documentation bugfixes

0.2_01    2015-01-03 16:41:07 Europe/Rome
   - added package for embedding one file in a Perl module
   - added sub-command in data-embed for feature above

0.1       2015-01-01 09:37:16 Europe/Rome
   - release after 71 passes on CPAN Testers (thanks!)

0.1_02    2014-12-31 14:09:57 Europe/Rome
   - workaround for perl bug in open() prior to version 5.14
   - added name() method to Data::Embed::File
   - added script for embedding/listing/extracting from container files

0.1_01    2014-12-30 17:47:32 Europe/Rome
   - developer release to see what CPAN Testers say


Makefile.PL

use strict;
use warnings;

use 5.008_000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "embed arbitrary data in a file",
  "AUTHOR" => "Flavio Poletti <polettix\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Data-Embed",
  "EXE_FILES" => [
    "bin/data-embed"
  ],
  "LICENSE" => "artistic_2",
  "NAME" => "Data::Embed",
  "PREREQ_PM" => {
    "IO::Slice" => "0.001",
    "Log::Log4perl::Tiny" => "1.002_004"
  },
  "TEST_REQUIRES" => {
    "Test::Exception" => "0.032",
    "Test::More" => "1.001008"
  },
  "VERSION" => "0.21",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "IO::Slice" => "0.001",
  "Log::Log4perl::Tiny" => "1.002_004",
  "Test::Exception" => "0.032",
  "Test::More" => "1.001008"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Data/Embed/File.pm
package Data::Embed::File;
{
  $Data::Embed::File::VERSION = '0.21';
}

# ABSTRACT: embed arbitrary data in a file

use strict;
use warnings;
use English qw< -no_match_vars >;
use IO::Slice;
use Fcntl qw< :seek >;
use Log::Log4perl::Tiny qw< :easy >;


sub new {
   my $package = shift;
   my $self = {(scalar(@_) && ref($_[0])) ? %{$_[0]} : @_};
   for my $feature (qw< offset length >) {
      LOGCROAK "$package new(): missing required field $feature"
        unless defined($self->{$feature})
        && $self->{$feature} =~ m{\A\d+\z}mxs;
   }
   LOGDIE "$package new(): either filename or fh are required"
     unless defined($self->{fh}) || defined($self->{filename});
   return bless $self, $package;
} ## end sub new


sub fh {
   my $self = shift;
   if (!exists $self->{slicefh}) {
      my %args = map { $_ => $self->{$_} }
        grep { defined $self->{$_} } qw< fh filename offset length >;
      $self->{slicefh} = IO::Slice->new(%args);
   }
   return $self->{slicefh};
} ## end sub fh


sub contents {
   my $self = shift;
   my $fh   = $self->fh();
   my $current = tell $fh;
   seek $fh, 0, SEEK_SET;

   local $/ = wantarray() ? $/ : undef;
   my @retval = <$fh>;
   seek $fh, $current, SEEK_SET;
   return @retval if wantarray();
   return $retval[0];
} ## end sub contents


sub name { return shift->{name}; }

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Data::Embed::File - embed arbitrary data in a file

=head1 VERSION

version 0.21

=head1 DESCRIPTION

Accessor class for representing an embedded file (for reading).

=head1 METHODS

=head2 B<< new >>

Constructor. It will act lazily, just storing the input data
for later usage by other methods, providing validation.

Input data can be provided as key-value pairs of through a
reference to a hash.

For proper functioning of the object, the following keys
should be provided:

=over

=item C<< fh >>

a filehandle for the stream where the data are contained

=item C<< filename >>

the name of the file where the data are. This parameter is
optional if C<fh> above is already provided.

=item C<< offset >>

the offset within the stream where the real data for this
file begins. C<0> means the very beginning of the file.

=item C<< length >>

the length of the data belonging to this C<File>.

=back

=head2 B<< fh >>

Get a filehandle suitable for accessing the embedded file. It provides
back a filehandle through L<IO::Slice>, providing the illusion of
working on a file per-se instead of a slice inside a bigger file.

=head2 B<< contents >>

Convenience method to slurp the whole contents of the embedded file
in one single shot. It always provides the full contents, independently
of whether data had been read before, although it restores the filehandle
to the previous position.

=head2 B<< name >>

Get the name associated to the file, whatever it is. L<Data::Embed::Reader>
sets it from what is read in the index file

=head1 AUTHOR

Flavio Poletti <polettix@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2015 by Flavio Poletti <polettix@cpan.org>

This module is free software.  You can redistribute it and/or
modify it under the terms of the Artistic License 2.0.

This program is distributed in the hope that it will be useful,
but without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

=cut


lib/Data/Embed/Util.pm
package Data::Embed::Util;
{
  $Data::Embed::Util::VERSION = '0.21';
}

# ABSTRACT: embed arbitrary data in a file - utilities

use strict;
use warnings;
use English qw< -no_match_vars >;

use Exporter qw< import >;
our @EXPORT_OK = qw< STARTER TERMINATOR escape unescape >;
our @EXPORT = (); # export nothing by default
our %EXPORT_TAGS = (
   all => \@EXPORT_OK,
   escaping => [qw< escape unescape >],
   constants => [qw< STARTER TERMINATOR >],
);


use constant STARTER    => "Data::Embed/index/begin\n";
use constant TERMINATOR => "Data::Embed/index/end\n";

sub escape {
   my $text = shift;
   $text =~ s{([^\w.-])}{'%' . sprintf('%02x', ord $1)}egmxs,
   return $text;
}

sub unescape {
   my $text = shift;
   $text =~ s{%(..)}{chr(hex($1))}egmxs;
   return $text;
}



1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Data::Embed::Util - embed arbitrary data in a file - utilities

=head1 VERSION

version 0.21

=head1 DESCRIPTION

Accessor class for representing an embedded file (for reading).

=head1 FUNCTIONS

All these functions are not exported by default. You can use the
following tags for importing them:

=over

=item C<< all >>

all of them

=item C<< escaping >>

the escaping functions C<escape> and C<unescape>

=item C<< constants >>

the two constants C<STARTER> and C<TERMINATOR>

=back

=head2 B<< STARTER >>

a string indicating the start of an index section

=head2 B<< TERMINATOR >>

a string indicating the end of an index section

=head2 B<< escape >>

escape an input string. All alphanumeric characters, plus the
underscore C<_>, hyphen C<-> and full stop C<.> are preserved, while
all the others are transformed into three-characters sequences C<%XX>
where XX are two hexadecimal digits.

=head2 B<< unescape >>

unescape an input string with the reverse action performed by
L</escape>.

=head1 AUTHOR

Flavio Poletti <polettix@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2015 by Flavio Poletti <polettix@cpan.org>

This module is free software.  You can redistribute it and/or
modify it under the terms of the Artistic License 2.0.

This program is distributed in the hope that it will be useful,
but without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

=cut


lib/Data/Embed/OneFileAsModule.pm
package Data::Embed::OneFileAsModule;
{
  $Data::Embed::OneFileAsModule::VERSION = '0.21';
}

# ABSTRACT: generate a Perl module for carrying data

use Exporter qw< import >;
@EXPORT_OK   = (qw< generate_module_from_file >);
@EXPORT      = ();
%EXPORT_TAGS = (all => \@EXPORT_OK);
use strict;
use warnings;
use Log::Log4perl::Tiny qw< :easy :dead_if_first >;



{
   no strict 'refs';
#__TEMPLATE_BEGIN__
{
   my $data     = \*{__PACKAGE__ . '::DATA'};
   my $position = undef;

   use strict;
   use warnings;
   use Carp;
   use English qw< -no_match_vars >;
   use Fcntl qw< :seek >;

   sub get_fh {
      $position = tell $data unless defined $position;
      open my $fh, '<&', $data
        or croak __PACKAGE__ . "::dup() for DATA: $OS_ERROR";
      seek $fh, $position, SEEK_SET;
      return $fh;
   } ## end sub get_fh

   sub get_data {
      my $fh = get_fh();
      binmode $fh;

      # ensure we slurp all, whatever the context
      local $/ = wantarray() ? $/ : undef;
      return <$fh>;
   } ## end sub get_data
}
#__TEMPLATE_END__
}

sub _get_output_fh {
   my $output = shift;

   # if no output is defined, we will return a scalar with data
   if (! defined $output) {
      my $buffer = '';
      open my $fh, '>', \$buffer
         or LOGCROAK "open() on (scalar ref): $OS_ERROR";
      return ($fh, \$buffer);
   }

   # if filename is '-', use standard output
   if ($output eq '-') {
      open my $fh, '>&', \*STDOUT        # dup()-ing
         or LOGCROAK "dup(): $OS_ERROR";
      binmode $fh;
      return $fh;
   }

   my $ro = ref $output;
   if (! $ro) { # output is a simple filename
      open my $fh, '>', $output
         or LOGCROAK "open('$output'): $OS_ERROR";
      binmode $fh;
      return $fh;
   }

   # so we have a reference here.
   # if not a reference to a SCALAR, assume it's
   # something that supports the filehandle interface
   return $output if $ro ne 'SCALAR';

   # otherwise, open a handle to write in the scalar
   open my $fh, '>', $output
      or LOGCROAK "open('$output') (scalar ref): $OS_ERROR";
   binmode $fh;
   return $fh;
}

sub _get_input_fh {
   my $args = shift;

   return $args->{fh} if exists $args->{fh};

   if (defined $args->{filename}) {
      open my $fh, '<', $args->{filename}
         or LOGCROAK "open('$args->{filename}'): $OS_ERROR";
      binmode $fh;
      return $fh;
   }

   if (defined $args->{dataref}) {
      open my $fh, '<', $args->{dataref}
         or LOGCROAK "open('$args->{dataref}') (scalar ref): $OS_ERROR";
      binmode $fh;
      return $fh;
   }

   if (defined $args->{data}) {
      open my $fh, '<', \$args->{data}
         or LOGCROAK "open() (scalar ref): $OS_ERROR";
      binmode $fh;
      return $fh;
   }

   LOGCROAK "no input source defined";
   return;    # unreached
}

sub generate_module_from_file {
   my %args = (scalar(@_) && ref($_[0])) ? %{$_[0]} : @_;

   LOGCROAK 'no package name set'
      unless defined $args{package};
   LOGCROAK "unsupported package name '$args{module}'"
      unless $args{package} =~ m{\A (?: \w+) (:: \w+)* \z}mxs;

   my $template_fh = get_fh();
   binmode $template_fh;
   seek $template_fh, 0, SEEK_SET;

   my $in_fh = _get_input_fh(\%args);

   ($args{output} = 'lib/' . $args{package} . '.pm') =~ s{::}{/}gmxs
      if $args{output_from_package};
   my ($out_fh, $outref) = _get_output_fh($args{output});

   # package name
   print {$out_fh} "package $args{package};\n";

   # package code
   my $seen_start;
   INPUT:
   while (<$template_fh>) {
      if (! $seen_start) {
         $seen_start = m{\A \#__TEMPLATE_BEGIN__ \s*\z}mxs;
         next INPUT;
      }
      last INPUT if m{\A \#__TEMPLATE_END__ \s*\z}mxs;
      print {$out_fh} $_;
   }

   # package code ending
   print {$out_fh} "\n1;\n__DATA__\n";

   # file contents
   while (! eof $in_fh) {
      defined(my $nread = read $in_fh, my $buffer, 4096)
         or LOGCROAK "read(): $OS_ERROR";
      last unless $nread; # paranoid
      print {$out_fh} $buffer;
   }

   return $$outref if $outref;
   return;
} ## end sub generate_module_from_file

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Data::Embed::OneFileAsModule - generate a Perl module for carrying data

=head1 VERSION

version 0.21

=head1 FUNCTIONS

=head2 get_fh

get a filehandle to read the data. The filehandle will be
put at the start of the data, you should not C<seek>
without taking into account that this is not at
position 0.

This function is preserved in the generated module so that
it is available to get the embedded data.

=head2 get_data

get a string with the full data provided by the carried by
the module.

This function is preserved in the generated module so that
it is available to get the embedded data.

=head2 generate_module_from_file

generate the data file contents. See full documentation
at L</Data::Embed::generate_module_from_file>.

=head1 AUTHOR

Flavio Poletti <polettix@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2015 by Flavio Poletti <polettix@cpan.org>

This module is free software.  You can redistribute it and/or
modify it under the terms of the Artistic License 2.0.

This program is distributed in the hope that it will be useful,
but without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

=cut


lib/Data/Embed.pm
package Data::Embed;
{
  $Data::Embed::VERSION = '0.21';
}

# ABSTRACT: embed arbitrary data in a file

use strict;
use warnings;
use English qw< -no_match_vars >;
use Exporter qw< import >;
use Log::Log4perl::Tiny qw< :easy :dead_if_first >;

our @EXPORT_OK =
  qw< writer reader embed embedded generate_module_from_file >;
our @EXPORT      = ();
our %EXPORT_TAGS = (
   all     => \@EXPORT_OK,
   reading => [qw< reader embedded >],
   writing => [qw< writer embed    generate_module_from_file >],
);


sub writer {
   require Data::Embed::Writer;
   return Data::Embed::Writer->new(@_);
}

sub reader {
   require Data::Embed::Reader;
   return Data::Embed::Reader->new(@_);
}

sub embed {
   my %args = (@_ && ref($_[0])) ? %{$_[0]} : @_;

   my %constructor_args =
     map { $_ => delete $args{$_} } qw< input output >;
   $constructor_args{input} = $constructor_args{output} =
     delete $args{container}
     if exists $args{container};
   my $writer = writer(%constructor_args)
     or LOGCROAK 'could not get the writer object';

   return $writer->add(%args);
} ## end sub embed

sub embedded {
   my $reader = reader(shift)
     or LOGCROAK 'could not get the writer object';
   return $reader->files();
}

sub generate_module_from_file {
   require Data::Embed::OneFileAsModule;
   goto &Data::Embed::OneFileAsModule::generate_module_from_file;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Data::Embed - embed arbitrary data in a file

=head1 VERSION

version 0.21

=head1 SYNOPSIS

   use Data::Embed qw< embed embedded >;

   # this is the file where thing will be embedded, at the end
   my $container = '/path/to/some/file';

   # first of all we embed an external file
   my $datafile  = '/path/to/data.tar.gz';
   embed($container, name => 'data.tar.gz', filename => $datafile);

   # we can also embed some data, directly
   use Data::Dumper;
   my $conf = { ... };
   embed($container, name => 'config.yml', data => Dumper($conf));

   # if the data is in a scalar but it's huge, use filename and
   # pass a reference to the scalar so no copy will happen
   my $huge_png = ...;
   embed($container, name => 'image.png', filename => \$huge_png);

   # to retrieve the stuff, use embedded()
   my @files = embedded($container);

   # each item in @files is a Data::Embed::File object

   # get whole contents of file
   my $config_text = $files[1]->contents();

   # otherwise, you can get a filehandle and use it, e.g. to
   # dump it on standard output
   my $data_fh = $files[0]->fh();
   binmode STDOUT;
   print {*STDOUT} <$data_fh>;

   # or save the file back, using the available name
   open my $ofh, '>:raw', $file[2]->name(); # well, do your checks!
   my $ifh = $files[2]->fh();
   while (! eof $ifh) {
      read $ifh, my $buffer, 4096
         or last; # do proper checks in production!
      print {$ofh} $buffer;
   }

=head1 DESCRIPTION

This module allows you to manage embedding data at the end of other
files, providing both means for embedding the data (L</embed> and
L</writer>) and accessing them (L</embedded> and L</reader>).

How can this be helpful? For example, suppose that you want to
bring some data along with your perl script, some of which might
be binary (e.g. an image, or a tar archive), you can embed these data
inside the perl and then retrieve them. For example, this can be the
basis for an installer.

For embedding data, you can use the L</embed> function. See the relevant
documentation or the examples in the L</SYNOPSYS> to use it properly.

For extracting the embedded data, you can use the L</embedded> function
and access each embedded file as a L<Data::Embed::File> object. You can
then use its methods C<contents> for accessing the whole data, or get
a filehandle through C<fh> and avoid getting the whole data in memory
at once.

Note: the filehandle provided by the C<fh> method of L<Data::Embed::File>
is actually a L<IO::Slice> object, so it might not support all the
functions/methods of a regular filehandle.

You can also access the lower level interface through the two functions
L</reader> and L</writer>. See the documentation
for L<Data::Embed::Reader> and L<Data::Embed::Writer>.

=head1 FUNCTIONS

=head2 B<< embed >>

   embed($hashref); # OR
   embed(%keyvalue_pairs);

Embed new data inside a container file.

Parameters can be passed as key-value pairs either
directly or through a hash reference. The following keys are
supported:

=over

=item C<container>

shortcut to specifying the same input and output, i.e. the value will
be replicated both on the C<input> and C<output> keys below. Caller
still has to ensure that the two are compatible. Provision of a
filehandle is currently not supported.

=item C<input>

any previous container file to use as base for the generated container.
If missing, no previous data will be considered (like starting from an
empty file). Can be:

=over

=item *

the C<-> string in a plain scalar, in which case standard input is
considered

=item *

any other string in a plain scalar, considered to be a file name

=item *

a plain reference to a scalar, considered to hold the input data

=item *

something that supports the filehandle interface for reading

=back

=item C<output>

the target container for the newly generated archive. Might be the same
as the input or different; in the latter case, the input will be copied
over the output, apart from the bits regarding the management of the
inclusions. Can be:

=over

=item *

missing/undefined or the C<-> string in a plain scalar, in which case
standard output is used

=item *

any other string in a plain scalar, considered to be a file name

=item *

a plain reference to a scalar, considered to be the target scalar to
hold the data

=item *

something that supports the filehandle interface for printing. You
should not provide the same filehandle for both input and output,
even if you opened it in read-write mode. This limitation might
be removed in the future.

=back

=item C<name>

the name to associate to the section, optionally. If missing it will
be set to the empty string

=item C<fh>

the filehandle from where data should be taken. The filehandle will be
exausted starting from its current position

=item C<filename>

a filename or a reference to a scalar where data will be read from

=item C<data>

a scalar from where data will be read. If you have a huge amount of
data, it's better to use the C<filename> key above passing a reference
to the scalar holding the data.

=back

Options C<fh>, C<filename> and C<data> are exclusive and will be considered
in the order above (first come, first served).

This function does not return anything.

=head2 B<< embedded >>

Get a list of the embedded files inside a target container. The calling
syntax is as follows:

   my $arrayref = embedded($container); # scalar context, OR
   my @files    = embedded($container); # list context

The only input parameter is the C<$container> to use as input. It can
be either a real filename, or a filehandle.

Depending on the context, a list will be returned (in list context) or
an array reference holding the list.

Whatever the context, each item in the list is a L<Data::Embed::File>
object that you can use to access the embedded file data (most notably,
you'll be probably using its C<contents> or C<fh> methods).

=head2 B<< writer >>

This is a convenience wrapper around the constructor for
L<Data::Embed::Writer>.

=head2 B<< reader >>

This is a convenience wrapper around the constructor for
L<Data::Embed::Reader>.

=head2 B<< generate_module_from_file >>

   # when %args includes details for an output channel
   generate_module_from_file(%args);

   # in case no output is provided in %args:
   my $text = generate_module_from_file(%args);

Generate a module's file contents from a file. The module contains code
of a package that has code to read the included data. Arguments are:

=over

=item package

the name of the package that will be put into the module. This
is a mandatory parameter.

=item output

the output channel. If not present, the output will be provided as
a string returned by the function, otherwise you can provide

=over

=item *

a filehandle where the output will be printed

=item *

a reference to a scalar (it will be filled with the contents)

=item *

the C<-> string, in which case the output will be printed
to STDOUT

=item *

a filename

=back

=item output_from_package

if this key is present and true, the C<output> parameters is
overridden and generated automatically from the package name
provided in key C<package>. The generated file will assume that
the file is contained in the I<normal> path under a C<lib>
directory, e.g. if the package name is C<Some::Module> then
the generated filename will be C<lib/Some/Module.pm>.

=item fh

a filehandle where data will be read from

=item filename

the input will be taken from the provided filename

=item dataref

the input will be taken from the scalar pointed by the
reference

=item data

the input is taken from the scalar provided with the data key

=back

Input keys are C<fh>, C<filename>, C<dataref> and C<data>. In case
multiple of them are present, they will be considered in the
order specified.

=head1 BUGS AND LIMITATIONS

Report bugs either through RT or GitHub (patches welcome).

Passing the same filehandle for both C<input> and C<output> in L</embed>
is not supported. This applies to C<container> too.

=head1 SEE ALSO

L<Data::Section> covers a somehow similar need but differently. In
particular, you should look at it if you want to be able to modify
the data you want to embed directly, e.g. if you are embedding some
textual templates that you want to tweak.

=head1 AUTHOR

Flavio Poletti <polettix@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2015 by Flavio Poletti <polettix@cpan.org>

This module is free software.  You can redistribute it and/or
modify it under the terms of the Artistic License 2.0.

This program is distributed in the hope that it will be useful,
but without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

=cut
